python/tvm/relay/frontend/tflite.py

```diff
@@ -0,0 +1,672 @@
+            raise ImportError("The tflite package must be installed")
+
+        assert isinstance(op, Operator)
+        input_tensors = self.get_input_tensors(op)
+        assert len(input_tensors) == 1, "input tensors length should be 1"
+
+        input_tensor = input_tensors[0]
+        input_tensor_idx = input_tensor.tensor_idx
+        params = {'axis': 1}  # 1 is channel
+        in_expr = self.get_expr(input_tensor_idx)
+        out = _op.nn.softmax(in_expr, **params)
+
+        return out
+
+    def convert_squeeze(self, op):
+        """Convert TFLite squeeze"""
+        try:
+            from tflite.BuiltinOptions import BuiltinOptions
+            from tflite.Operator import Operator
+            from tflite.SqueezeOptions import SqueezeOptions
+        except ImportError:
+            raise ImportError("The tflite package must be installed")
+
+        assert isinstance(op, Operator)
+        input_tensors = self.get_input_tensors(op)
+        output_tensors = self.get_output_tensors(op)
+        assert len(input_tensors) == 1, "input tensors length should be 1"
+        assert len(output_tensors) == 1, "output tensors length should be 1"
+        input_tensor = input_tensors[0]
+        input_tensor_idx = input_tensor.tensor_idx
+
+        assert op.BuiltinOptionsType() == BuiltinOptions.SqueezeOptions
+        op_options = op.BuiltinOptions()
+        squeeze_options = SqueezeOptions()
+        squeeze_options.Init(op_options.Bytes, op_options.Pos)
+        squeeze_axis = squeeze_options.SqueezeDimsAsNumpy()
+        input_shape_length = len(input_tensor.tensor.ShapeAsNumpy())
+        output_shape_length = len(output_tensors[0].tensor.ShapeAsNumpy())
+
+        in_expr = self.get_expr(input_tensor_idx)
+
+        # TFLite is N H W C, our layout is N C H W
+        if input_shape_length == 1 or input_shape_length == 2:
+            # The rule is channel first (after N but before H, W).
+            # length of 1 means N*H*W*C, do nothing.
+            # length of 2 means N*H*W, C, do nothing.
+            pass
+        elif input_shape_length == 3:
+            # convert N C H*W to N H*W C
+            in_expr = _op.transpose(in_expr, axes=(0, 2, 1))
+        elif input_shape_length == 4:
+            # convert input to N H W C, then reshape to target shape,
+            # finally convert back if necessary
+            in_expr = _op.transpose(in_expr, axes=(0, 2, 3, 1))
+        else:
+            raise NotImplementedError("Not support input shape length {} of squeeze : "
+                                      .format(str(input_shape_length)))
+
+        out = _op.squeeze(in_expr, axis=tuple(squeeze_axis))
+
+        # The rule is channel first.
+        # 1: N*H*W*C
+        # 2: N*H*W, C
+        # 3: N H W C, reshape to N H*W C, transpose to N C H*W
+        # 4: N H W C, transpose to N C H W
+        # add more if we need target shapes in future
+        if output_shape_length == 1 or output_shape_length == 2:
```

Comment:
Ok, I understand the logic here, but I do think it introduces some underlying assumption.

The assumption is: the input of squeeze operator comes out of conv2d with some specific layout, otherwise it will transpose layers crazily.

It doesn't look safe to me, but for now for specific models like MobileNet, it is fine. @tqchen please advise.

Reply:
Yes. When we do handle `Reshape` or `squeeze` ops we should be very careful if we handle from input layout NHWC to NCHW. So, this is to obey channel first rule, i.e. C is before H, W but after N. Currently we only handle 1-4D, not N-D. This issue is not special for us, other converters will meet the same issue. Please see Tensorflow to CoreML converter, which has similar logic: https://github.com/tf-coreml/tf-coreml/blob/master/tfcoreml/_shape_sensitive_layers.py#L275

Reply:
I feel like it is ok for me for now, but probably we could document it somewhere? DL models in other fields like NLP heavily use squeeze, which may violate our rules and crashes every thing.